File: tests/unit/orbit_generation/eclipse/test_compute_eclipse_cylinder.py
import sys
from pathlib import Path
import numpy as np
import pytest
import math

ROOT = Path(__file__).resolve().parents[4]
sys.path.insert(0, str(ROOT / "build"))

import saltro_py

# Constants
RE = 6378136.3  # Earth's radius in meters
AU_M = 1.496e11  # AU in meters


class TestEclipseCylinderBasicFunctionality:
    """Test basic eclipse cylinder operation"""
    
    def test_basic_functionality(self):
        """Sunlit positions should remain unchanged"""
        N = 5
        R = np.zeros((3, N))
        S = np.zeros((3, N))
        
        for i in range(N):
            R[:, i] = [RE + 400e3, 0.0, 0.0]
            S[:, i] = [1e11, 0.0, 0.0]
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        # All positions should be sunlit
        for i in range(N):
            sun_norm = np.linalg.norm(S_out[:, i])
            assert sun_norm > 1e10
    
    def test_detects_shadow(self):
        """Satellite on night side should be eclipsed"""
        R = np.array([[-RE - 400e3, 0.0, 0.0]])
        S = np.array([[1e11, 0.0, 0.0]])
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R.T, S.T)
        assert ok
        assert np.linalg.norm(S_out[:, 0]) < 1e-10
    
    def test_sunlit_side_unchanged(self):
        """Sunlit positions should preserve sun vector"""
        R = np.array([RE + 400e3, 0.0, 0.0]).reshape(3, 1)
        S_orig = np.array([1e11, 0.0, 0.0]).reshape(3, 1)
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S_orig)
        assert ok
        assert np.linalg.norm(S_out[:, 0] - S_orig[:, 0]) < 1e-6


class TestEclipseCylinderGeometry:
    """Test geometric edge cases"""
    
    def test_off_axis_night_side(self):
        """Satellite behind Earth but offset should still be eclipsed"""
        R = np.array([-(RE + 400e3), RE * 0.5, 0.0]).reshape(3, 1)
        S = np.array([1e11, 0.0, 0.0]).reshape(3, 1)
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        assert np.linalg.norm(S_out[:, 0]) < 1e-10
    
    def test_beyond_shadow_edge(self):
        """Satellite beyond shadow edge should be sunlit"""
        R = np.array([-(RE + 400e3), RE * 2.0, 0.0]).reshape(3, 1)
        S_orig = np.array([1e11, 0.0, 0.0]).reshape(3, 1)
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S_orig)
        assert ok
        assert np.linalg.norm(S_out[:, 0] - S_orig[:, 0]) < 1e-6


class TestEclipseCylinderOrbit:
    """Test eclipse patterns for different orbits"""
    
    def test_circular_orbit_partial_eclipse(self):
        """Circular orbit should have partial eclipse"""
        N = 24
        R = np.zeros((3, N))
        S = np.zeros((3, N))
        
        alt = 400e3
        r = RE + alt
        
        for i in range(N):
            theta = 2.0 * math.pi * i / N
            R[:, i] = [r * math.cos(theta), r * math.sin(theta), 0.0]
            S[:, i] = [AU_M, 0.0, 0.0]
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        eclipsed_count = sum(1 for i in range(N) if np.linalg.norm(S_out[:, i]) < 1e-10)
        sunlit_count = N - eclipsed_count
        
        assert eclipsed_count > 0
        assert sunlit_count > 0
    
    def test_equatorial_orbit(self):
        """Equatorial orbit should have binary eclipse/sunlit"""
        N = 16
        R = np.zeros((3, N))
        S = np.zeros((3, N))
        
        r = RE + 400e3
        
        for i in range(N):
            theta = 2.0 * math.pi * i / N
            R[:, i] = [r * math.cos(theta), r * math.sin(theta), 0.0]
            S[:, i] = [AU_M, 0.0, 0.0]
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        for i in range(N):
            sun_mag = np.linalg.norm(S_out[:, i])
            is_sunlit = sun_mag > 1e10
            is_eclipsed = sun_mag < 1e-10
            assert is_sunlit or is_eclipsed
    
    def test_high_altitude_orbit(self):
        """GEO altitude should have minimal eclipse"""
        N = 12
        R = np.zeros((3, N))
        S = np.zeros((3, N))
        
        r = RE + 36000e3  # GEO
        
        for i in range(N):
            theta = 2.0 * math.pi * i / N
            R[:, i] = [r * math.cos(theta), r * math.sin(theta), 0.0]
            S[:, i] = [AU_M, 0.0, 0.0]
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        eclipsed_count = sum(1 for i in range(N) if np.linalg.norm(S_out[:, i]) < 1e-10)
        assert eclipsed_count < N // 4  # Less than 25%


class TestEclipseCylinderVectorPreservation:
    """Test that sun vectors are preserved when appropriate"""
    
    def test_preserves_magnitude_sunlit(self):
        """Sunlit positions should preserve vector magnitude"""
        N = 3
        R = np.zeros((3, N))
        S = np.zeros((3, N))
        
        for i in range(N):
            R[:, i] = [RE + 400e3 + i * 100e3, 0.0, 0.0]
            S[:, i] = [AU_M, 1e9, 1e8]
        
        original_mags = np.array([np.linalg.norm(S[:, i]) for i in range(N)])
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        for i in range(N):
            new_mag = np.linalg.norm(S_out[:, i])
            assert abs(new_mag - original_mags[i]) < 1e-6
    
    def test_zero_sun_vector_handling(self):
        """Zero sun vectors should be skipped"""
        R = np.zeros((3, 2))
        S = np.zeros((3, 2))
        
        R[:, 0] = [RE + 400e3, 0.0, 0.0]
        S[:, 0] = [1e11, 0.0, 0.0]
        
        R[:, 1] = [-(RE + 400e3), 0.0, 0.0]
        S[:, 1] = [0.0, 0.0, 0.0]
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        assert np.linalg.norm(S_out[:, 0]) > 1e10
        assert np.linalg.norm(S_out[:, 1]) < 1e-10


class TestEclipseCylinderMathematical:
    """Test mathematical correctness of eclipse calculations"""
    
    def test_multi_orbit_mathematical_model(self):
        """Multi-orbit test with explicit mathematical calculation"""
        N = 100  # 5 complete orbits of 20 points each
        R = np.zeros((3, N))
        S = np.zeros((3, N))
        
        r = RE + 400e3
        
        for i in range(N):
            theta = 2.0 * math.pi * i / (N / 5)  # 5 orbits
            R[:, i] = [r * math.cos(theta), r * math.sin(theta), 0.0]
            S[:, i] = [AU_M, 0.0, 0.0]
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        eclipsed_per_orbit = sum(1 for i in range(N) if np.linalg.norm(S_out[:, i]) < 1e-10)
        
        # Calculate expected eclipse fraction from geometry
        # For circular orbit at altitude h: sin(θ_e) = R_E / (R_E + h)
        # where θ_e is the eclipse half-angle
        alt = 400e3
        eclipse_half_angle = math.asin(RE / (RE + alt))
        total_eclipse_angle = 2.0 * eclipse_half_angle  # in radians
        eclipse_fraction = total_eclipse_angle / (2.0 * math.pi)
        
        # Expected eclipsed count with ±10% tolerance for discretization
        expected_eclipsed = int(eclipse_fraction * N)
        lower_bound = int(expected_eclipsed * 0.9)
        upper_bound = int(expected_eclipsed * 1.1)
        
        assert eclipsed_per_orbit >= lower_bound
        assert eclipsed_per_orbit <= upper_bound
    
    def test_arbitrary_sun_direction(self):
        """Test with non-axis-aligned sun direction"""
        N = 8
        R = np.zeros((3, N))
        S = np.zeros((3, N))
        
        r = RE + 400e3
        sun_dir = np.array([1.0, 0.5, 0.3])
        sun_dir = sun_dir / np.linalg.norm(sun_dir)
        sun_dist = AU_M
        
        for i in range(N):
            theta = 2.0 * math.pi * i / N
            R[:, i] = [r * math.cos(theta), r * math.sin(theta), 0.0]
            S[:, i] = sun_dist * sun_dir
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        
        # Verify all outputs are either normal or zero
        for i in range(N):
            sun_mag = np.linalg.norm(S_out[:, i])
            is_normal = abs(sun_mag - sun_dist) < 1e4
            is_zero = sun_mag < 1e-10
            assert is_normal or is_zero


class TestEclipseCylinderEdgeCases:
    """Test edge cases and boundary conditions"""
    
    def test_single_point(self):
        """Single sunlit point"""
        R = np.array([RE + 400e3, 0.0, 0.0]).reshape(3, 1)
        S = np.array([AU_M, 0.0, 0.0]).reshape(3, 1)
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        assert np.linalg.norm(S_out[:, 0]) > 1e10
    
    def test_directly_behind_earth(self):
        """Satellite directly behind Earth should be eclipsed"""
        R = np.array([-(RE + 400e3), 0.0, 0.0]).reshape(3, 1)
        S = np.array([AU_M, 0.0, 0.0]).reshape(3, 1)
        
        ok, S_out = saltro_py.compute_eclipse_cylinder(R, S)
        assert ok
        assert np.linalg.norm(S_out[:, 0]) < 1e-10


if __name__ == "__main__":
    pytest.main([__file__, "-v"])

File: tests/unit/orbit_generation/eclipse/test_compute_eclipse_cylinder.cpp
#include <catch2/catch_test_macros.hpp>
#include <Eigen/Dense>
#include <saltro/orbit_generation/eclipse/compute_eclipse_cylinder.h>
#include <saltro/limits.h>
#include <saltro/constants/constants.h>
#include <cmath>

using namespace saltro;

TEST_CASE("eclipse cylinder basic functionality", "[eclipse][cylinder]")
{
    constexpr int N = 5;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    for (int i = 0; i < N; ++i) {
        R.col(i) << Re + 400e3, 0.0, 0.0;  // Position above Earth
        S.col(i) << 1e11, 0.0, 0.0;        // Sun far away in +X direction
    }

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // All positions should be sunlit (positive dot product with sun direction)
    for (int i = 0; i < N; ++i) {
        double sun_norm = S.col(i).norm();
        REQUIRE(sun_norm > 1e10);  // Sun vector unchanged
    }
}


TEST_CASE("eclipse cylinder detects shadow", "[eclipse][cylinder]")
{
    constexpr int N = 1;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    // Satellite on night side (behind Earth relative to sun)
    R.col(0) << -(Re + 400e3), 0.0, 0.0;
    // Sun is infinitely far in +X direction
    S.col(0) << 1e11, 0.0, 0.0;

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Should be in eclipse (sun vector zeroed)
    REQUIRE(S.col(0).norm() < 1e-10);
}


TEST_CASE("eclipse cylinder sunlit side unchanged", "[eclipse][cylinder]")
{
    constexpr int N = 1;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    // Satellite on sunlit side
    R.col(0) << Re + 400e3, 0.0, 0.0;
    Eigen::Vector3d sun_original(1e11, 0.0, 0.0);
    S.col(0) = sun_original;

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Sun vector should be unchanged
    REQUIRE((S.col(0) - sun_original).norm() < 1e-6);
}


TEST_CASE("eclipse cylinder off-axis night side", "[eclipse][cylinder]")
{
    constexpr int N = 1;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    // Satellite behind Earth but off-axis (not directly behind)
    // Position: some distance behind Earth, offset perpendicular to sun direction
    R.col(0) << -(Re + 400e3), Re * 0.5, 0.0;
    S.col(0) << 1e11, 0.0, 0.0;

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Should still be in eclipse (within shadow cylinder)
    REQUIRE(S.col(0).norm() < 1e-10);
}


TEST_CASE("eclipse cylinder beyond shadow edge", "[eclipse][cylinder]")
{
    constexpr int N = 1;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    // Satellite behind Earth but far enough off-axis to be outside shadow
    R.col(0) << -(Re + 400e3), Re * 2.0, 0.0;
    Eigen::Vector3d sun_original(1e11, 0.0, 0.0);
    S.col(0) = sun_original;

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Outside shadow, sun vector should be unchanged
    REQUIRE((S.col(0) - sun_original).norm() < 1e-6);
}


TEST_CASE("eclipse cylinder circular orbit partial eclipse", "[eclipse][cylinder]")
{
    constexpr int N = 24;  // One full orbit

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;
    const double alt = 400e3;
    const double r = Re + alt;

    for (int i = 0; i < N; ++i) {
        double theta = 2.0 * M_PI * i / N;
        
        // Circular orbit in XY plane
        R.col(i) << r * std::cos(theta), r * std::sin(theta), 0.0;
        
        // Sun along +X direction (infinitely far)
        S.col(i) << 1.496e11, 0.0, 0.0;
    }

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Count eclipsed and sunlit positions
    int eclipsed_count = 0;
    int sunlit_count = 0;

    for (int i = 0; i < N; ++i) {
        if (S.col(i).norm() < 1e-10) {
            eclipsed_count++;
        } else {
            sunlit_count++;
        }
    }

    // Should have roughly half in eclipse, half sunlit
    REQUIRE(eclipsed_count > 0);
    REQUIRE(sunlit_count > 0);
    REQUIRE(eclipsed_count + sunlit_count == N);
}


TEST_CASE("eclipse cylinder preserves vector magnitude when sunlit", "[eclipse][cylinder]")
{
    constexpr int N = 3;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    for (int i = 0; i < N; ++i) {
        R.col(i) << Re + 400e3 + i * 100e3, 0.0, 0.0;  // Different altitudes
        S.col(i) << 1.496e11, 1e9, 1e8;                // Sun vector (not aligned with position)
    }

    // Store original magnitudes
    std::vector<double> original_mags(N);
    for (int i = 0; i < N; ++i) {
        original_mags[i] = S.col(i).norm();
    }

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // All should be sunlit, magnitudes unchanged
    for (int i = 0; i < N; ++i) {
        double new_mag = S.col(i).norm();
        REQUIRE(std::abs(new_mag - original_mags[i]) < 1e-6);
    }
}


TEST_CASE("eclipse cylinder zero sun vector handling", "[eclipse][cylinder]")
{
    constexpr int N = 2;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    R.col(0) << Re + 400e3, 0.0, 0.0;
    S.col(0) << 1e11, 0.0, 0.0;

    R.col(1) << -(Re + 400e3), 0.0, 0.0;
    S.col(1) << 0.0, 0.0, 0.0;  // Already zero sun vector

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // First should be sunlit
    REQUIRE(S.col(0).norm() > 1e10);

    // Second should remain zero (skipped)
    REQUIRE(S.col(1).norm() < 1e-10);
}


TEST_CASE("eclipse cylinder equatorial orbit", "[eclipse][cylinder]")
{
    constexpr int N = 16;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;
    const double r = Re + 400e3;

    // Orbit along equator
    for (int i = 0; i < N; ++i) {
        double theta = 2.0 * M_PI * i / N;
        R.col(i) << r * std::cos(theta), r * std::sin(theta), 0.0;
        S.col(i) << 1.496e11, 0.0, 0.0;  // Sun along +X
    }

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Check that eclipse transitions are smooth
    // (sun vector either nonzero or zero, no partial values for cylindrical model)
    for (int i = 0; i < N; ++i) {
        double sun_mag = S.col(i).norm();
        bool is_sunlit = sun_mag > 1e10;
        bool is_eclipsed = sun_mag < 1e-10;
        REQUIRE((is_sunlit || is_eclipsed));
    }
}


TEST_CASE("eclipse cylinder high altitude orbit", "[eclipse][cylinder]")
{
    constexpr int N = 12;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;
    const double r = Re + 36000e3;  // GEO altitude

    for (int i = 0; i < N; ++i) {
        double theta = 2.0 * M_PI * i / N;
        R.col(i) << r * std::cos(theta), r * std::sin(theta), 0.0;
        S.col(i) << 1.496e11, 0.0, 0.0;
    }

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // At GEO altitude, eclipse window is very small
    int eclipsed_count = 0;
    for (int i = 0; i < N; ++i) {
        if (S.col(i).norm() < 1e-10) {
            eclipsed_count++;
        }
    }

    // Should have some eclipse but much less than at LEO
    REQUIRE(eclipsed_count >= 0);
    REQUIRE(eclipsed_count < N / 4);  // Less than 25% in eclipse
}


TEST_CASE("eclipse cylinder single point", "[eclipse][cylinder]")
{
    constexpr int N = 1;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;

    R.col(0) << Re + 400e3, 0.0, 0.0;
    S.col(0) << 1.496e11, 0.0, 0.0;

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Should be sunlit
    REQUIRE(S.col(0).norm() > 1e10);
}


TEST_CASE("eclipse cylinder multi-orbit degradation", "[eclipse][cylinder]")
{
    constexpr int N = 100;  // Multiple orbits

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;
    const double r = Re + 400e3;

    for (int i = 0; i < N; ++i) {
        double theta = 2.0 * M_PI * i / (N / 5);  // 5 full orbits
        R.col(i) << r * std::cos(theta), r * std::sin(theta), 0.0;
        S.col(i) << 1.496e11, 0.0, 0.0;
    }

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Verify consistent eclipse/sunlit pattern across all orbits
    int eclipsed_per_orbit = 0;
    for (int i = 0; i < N; ++i) {
        if (S.col(i).norm() < 1e-10) {
            eclipsed_per_orbit++;
        }
    }

    // Calculate expected eclipse fraction from geometry
    // For circular orbit at altitude h: sin(θ_e) = R_E / (R_E + h)
    // where θ_e is the eclipse half-angle
    const double alt = 400e3;
    const double eclipse_half_angle = std::asin(Re / (Re + alt));
    const double total_eclipse_angle = 2.0 * eclipse_half_angle;  // in radians
    const double eclipse_fraction = total_eclipse_angle / (2.0 * M_PI);
    
    // Expected eclipsed count with ±10% tolerance for discretization
    const int expected_eclipsed = static_cast<int>(eclipse_fraction * N);
    const int lower_bound = static_cast<int>(expected_eclipsed * 0.9);
    const int upper_bound = static_cast<int>(expected_eclipsed * 1.1);
    
    REQUIRE(eclipsed_per_orbit >= lower_bound);
    REQUIRE(eclipsed_per_orbit <= upper_bound);
}


TEST_CASE("eclipse cylinder arbitrary sun direction", "[eclipse][cylinder]")
{
    constexpr int N = 8;

    Eigen::Matrix<double, 3, limits::MAX_LENGTH_TRAJ> R, S;
    const double Re = saltro::constants::R_EARTH;
    const double r = Re + 400e3;

    // Sun direction: not aligned with any axis
    Eigen::Vector3d sun_dir(1.0, 0.5, 0.3);
    sun_dir.normalize();
    double sun_dist = 1.496e11;

    for (int i = 0; i < N; ++i) {
        double theta = 2.0 * M_PI * i / N;
        R.col(i) << r * std::cos(theta), r * std::sin(theta), 0.0;
        S.col(i) = sun_dist * sun_dir;
    }

    bool ok = orbits::compute_eclipse_cylinder(R, N, S);
    REQUIRE(ok);

    // Verify all outputs are either normal or zero
    for (int i = 0; i < N; ++i) {
        double sun_mag = S.col(i).norm();
        bool is_normal = std::abs(sun_mag - sun_dist) < 1e4;
        bool is_zero = sun_mag < 1e-10;
        REQUIRE((is_normal || is_zero));
    }
}
